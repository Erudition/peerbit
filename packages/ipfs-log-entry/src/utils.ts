export const bigIntMax = (...args) => args.reduce((m, e) => e > m ? e : m);
export const bigIntMin = (...args) => args.reduce((m, e) => e < m ? e : m);
