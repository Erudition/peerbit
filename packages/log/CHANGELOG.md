## @dao-xyz/peerbit-log [0.8.6](https://github.com/dao-xyz/peerbit/compare/@dao-xyz/peerbit-log@0.8.5...@dao-xyz/peerbit-log@0.8.6) (2023-06-07)


### Bug Fixes

* add release cfg ([de76654](https://github.com/dao-xyz/peerbit/commit/de766548f8106804d319e8b51e9607f2a3f60726))





### Dependencies

* **@dao-xyz/cache:** upgraded to 0.0.6
* **@dao-xyz/libp2p-direct-block:** upgraded to 0.5.3
* **@dao-xyz/peerbit-crypto:** upgraded to 0.1.16
* **@dao-xyz/peerbit-logger:** upgraded to 0.0.7
* **@dao-xyz/peerbit-keystore:** upgraded to 0.2.12
* **@dao-xyz/peerbit-test-utils:** upgraded to 0.4.3
