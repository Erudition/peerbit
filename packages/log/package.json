{
    "name": "@dao-xyz/peerbit-log",
    "version": "0.0.9",
    "description": "Append-only log CRDT",
    "author": "dao.xyz",
    "license": "MIT",
    "type": "module",
    "module": "lib/esm/index.js",
    "types": "lib/esm/index.d.ts",
    "exports": {
        "import": "./lib/esm/index.js",
        "require": "./lib/cjs/index.js"
    },
    "files": [
        "lib",
        "src",
        "!src/**/__tests__",
        "!lib/**/__tests__",
        "LICENSE"
    ],
    "publishConfig": {
        "access": "public"
    },
    "engines": {
        "node": ">=16.15.1"
    },
    "keywords": [
        "ipfs",
        "log",
        "crdts",
        "crdt"
    ],
    "dependencies": {
        "@dao-xyz/borsh": "^4.0.7",
        "@dao-xyz/peerbit-block": "^0.0.13",
        "@dao-xyz/peerbit-crypto": "^0.0.33",
        "@dao-xyz/peerbit-logger": "^0.0.4",
        "json-stringify-deterministic": "^1.0.7",
        "multihashing-async": "^2.0.1",
        "p-do-whilst": "^1.1.0",
        "p-each-series": "^3.0.0",
        "p-map": "^5.5.0",
        "p-whilst": "^3.0.0",
        "yallist": "^4.0.0"
    },
    "devDependencies": {
        "@dao-xyz/peerbit-keystore": "^0.0.45",
        "@dao-xyz/peerbit-test-utils": "^0.0.24",
        "@types/yallist": "^4.0.1",
        "assert": "^2.0.0",
        "fs-extra": "^10.0.1",
        "is-node": "^1.0.2",
        "path-browserify": "^1.0.1",
        "stream-browserify": "^3.0.0"
    },
    "contributors": [
        "haadcode",
        "aphelionz",
        "shamb0t",
        "thiagodelgado111",
        "mistakia",
        "satazor",
        "RichardLitt",
        "greenkeeperio-bot",
        "chrisdostert",
        "zachferland",
        "kaibakker",
        "dignifiedquire",
        "adam-palazzo"
    ],
    "scripts": {
        "clean": "shx rm -rf lib/*",
        "build": "yarn clean && tsc -p tsconfig.json",
        "postbuild": "echo '{\"type\":\"module\"} ' | node ../../node_modules/.bin/json > lib/esm/package.json",
        "test": "node --experimental-vm-modules ./../../node_modules/.bin/jest test  -c  ../../jest.config.ts --runInBand --forceExit",
        "test:unit": "node --experimental-vm-modules ../../node_modules/.bin/jest test  -c  ../../jest.config.unit.ts --runInBand --forceExit",
        "test:integration": "node --experimental-vm-modules ../node_modules/.bin/jest test -c  ../../jest.config.integration.ts --runInBand --forceExit",
        "benchmark": "nyc  --require ts-node/register benchmark-runner -r -b --baselineLimit 1000",
        "benchmark:stress": "benchmark-runner -r --grep stress"
    },
    "localMaintainers": [
        "dao.xyz"
    ],
    "gitHead": "bfbec720aee276ab1bc7cf4ee7a6c0e90ecb4f51"
}
