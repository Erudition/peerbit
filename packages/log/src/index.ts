export * from "./log.js";
export * from "./log-sorting.js";
export * from "./log-errors.js";
export * from "./log-io.js";
export * from "./entry.js";
export * from "./utils.js";
export * from "./clock.js";
export * from "./encoding.js";
export * from "./identity.js";
export type { TrimToByteLengthOption, TrimToLengthOption } from "./trim.js";
