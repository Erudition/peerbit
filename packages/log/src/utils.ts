export const max = <T>(...args: T[]) => args.reduce((m, e) => (e > m ? e : m));
export const min = <T>(...args: T[]) => args.reduce((m, e) => (e < m ? e : m));
