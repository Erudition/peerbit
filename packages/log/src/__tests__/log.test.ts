import assert from "assert";
import rmrf from "rimraf";
import { Entry, Payload } from "../entry.js";
import { LamportClock as Clock, Timestamp } from "../clock.js";
import { Log } from "../log.js";
import { Keystore, KeyWithMeta } from "@dao-xyz/peerbit-keystore";
import fs from "fs-extra";
import {
	BlockStore,
	MemoryLevelBlockStore,
} from "@dao-xyz/libp2p-direct-block";
import { createBlock, getBlockValue } from "@dao-xyz/libp2p-direct-block";

import { LastWriteWins } from "../log-sorting.js";
import { signingKeysFixturesPath, testKeyStorePath } from "./utils.js";
import { Ed25519Keypair } from "@dao-xyz/peerbit-crypto";
import { dirname } from "path";
import { fileURLToPath } from "url";
import path from "path";
import { compare } from "@dao-xyz/uint8arrays";
import { createStore } from "./utils.js";

const __filename = fileURLToPath(import.meta.url);
const __filenameBase = path.parse(__filename).base;
const __dirname = dirname(__filename);

// For tiebreaker testing
const FirstWriteWins = (a: any, b: any) => LastWriteWins(a, b) * -1;

let signKey: KeyWithMeta<Ed25519Keypair>,
	signKey2: KeyWithMeta<Ed25519Keypair>,
	signKey3: KeyWithMeta<Ed25519Keypair>;

describe("Log", function () {
	let keystore: Keystore;
	let store: BlockStore;
	beforeAll(async () => {
		await fs.copy(
			signingKeysFixturesPath(__dirname),
			testKeyStorePath(__filenameBase)
		);

		keystore = new Keystore(
			await createStore(testKeyStorePath(__filenameBase))
		);
		const signKeys: KeyWithMeta<Ed25519Keypair>[] = [];
		for (let i = 0; i < 3; i++) {
			signKeys.push(
				(await keystore.getKey(
					new Uint8Array([i])
				)) as KeyWithMeta<Ed25519Keypair>
			);
		}
		signKeys.sort((a, b) =>
			compare(a.keypair.publicKey.publicKey, b.keypair.publicKey.publicKey)
		);
		// @ts-ignore
		signKey = signKeys[0];
		// @ts-ignore
		signKey2 = signKeys[1];
		// @ts-ignore
		signKey3 = signKeys[2];
		store = new MemoryLevelBlockStore();
		await store.open();
	});

	afterAll(async () => {
		await store.close();
		rmrf.sync(testKeyStorePath(__filenameBase));

		await keystore?.close();
	});

	describe("constructor", () => {
		it("creates an empty log with default params", async () => {
			const log = new Log();
			await log.open(
				store,
				{
					...signKey.keypair,
					sign: async (data: Uint8Array) => await signKey.keypair.sign(data),
				},
				undefined
			);
			assert.notStrictEqual(log.entryIndex, null);
			assert.notStrictEqual(log.headsIndex, null);
			assert.notStrictEqual(log.id, null);
			assert.notStrictEqual(log.id, null);
			assert.notStrictEqual(log.toArray(), null);
			assert.notStrictEqual(await log.getHeads(), null);
			assert.deepStrictEqual(await log.toArray(), []);
			assert.deepStrictEqual(await log.getHeads(), []);
			assert.deepStrictEqual(await log.getTails(), []);
			assert.deepStrictEqual(await log.getTailHashes(), []);
		});

		it("can not setup after open", async () => {
			const log = new Log();
			await log.open(
				store,
				{
					...signKey.keypair,
					sign: async (data: Uint8Array) => await signKey.keypair.sign(data),
				},
				undefined
			);
			await expect(() => log.setup()).rejects.toThrow();
		});
		it("sets an id", async () => {
			const log = new Log({ id: new Uint8Array(1) });
			await log.open(store, {
				...signKey.keypair,
				sign: async (data: Uint8Array) => await signKey.keypair.sign(data),
			});
			expect(log.id).toEqual(new Uint8Array(1));
		});

		it("generates if id is not passed as an argument", async () => {
			const log = new Log();
			await log.open(store, {
				...signKey.keypair,
				sign: async (data: Uint8Array) => await signKey.keypair.sign(data),
			});
			expect(log.id).toBeInstanceOf(Uint8Array);
		});

		it("sets items if given as params", async () => {
			const one = await Entry.create({
				store,
				identity: {
					...signKey.keypair,
					sign: (data) => signKey.keypair.sign(data),
				},
				gidSeed: Buffer.from("a"),
				data: "entryA",
				next: [],
				clock: new Clock({ id: new Uint8Array([0]), timestamp: 0 }),
			});
			const two = await Entry.create({
				store,
				identity: {
					...signKey.keypair,
					sign: (data) => signKey.keypair.sign(data),
				},
				gidSeed: Buffer.from("a"),
				data: "entryB",
				next: [],
				clock: new Clock({ id: new Uint8Array([1]), timestamp: 0 }),
			});
			const three = await Entry.create({
				store,
				identity: {
					...signKey.keypair,
					sign: (data) => signKey.keypair.sign(data),
				},
				gidSeed: Buffer.from("a"),
				data: "entryC",
				next: [],
				clock: new Clock({ id: new Uint8Array([2]), timestamp: 0 }),
			});
			const log = new Log<string>();
			await log.open(store, {
				...signKey.keypair,
				sign: async (data: Uint8Array) => await signKey.keypair.sign(data),
			});
			await log.reset([one, two, three]);

			expect(log.length).toEqual(3);
			expect((await log.toArray())[0].payload.getValue()).toEqual("entryA");
			expect((await log.toArray())[1].payload.getValue()).toEqual("entryB");
			expect((await log.toArray())[2].payload.getValue()).toEqual("entryC");
		});

		it("sets heads if given as params", async () => {
			const one = await Entry.create({
				store,
				identity: {
					...signKey.keypair,
					sign: (data) => signKey.keypair.sign(data),
				},
				gidSeed: Buffer.from("a"),
				data: "entryA",
				next: [],
			});
			const two = await Entry.create({
				store,
				identity: {
					...signKey.keypair,
					sign: (data) => signKey.keypair.sign(data),
				},
				gidSeed: Buffer.from("a"),
				data: "entryB",
				next: [],
			});
			const three = await Entry.create({
				store,
				identity: {
					...signKey.keypair,
					sign: (data) => signKey.keypair.sign(data),
				},
				gidSeed: Buffer.from("a"),
				data: "entryC",
				next: [],
			});
			const log = new Log<string>();
			await log.open(store, {
				...signKey.keypair,
				sign: async (data: Uint8Array) => await signKey.keypair.sign(data),
			});
			await log.reset([one, two, three], [three]);

			expect((await log.getHeads()).length).toEqual(1);
			expect((await log.getHeads())[0].hash).toEqual(three.hash);
		});

		it("finds heads if heads not given as params", async () => {
			const one = await Entry.create({
				store,
				identity: {
					...signKey.keypair,
					sign: (data) => signKey.keypair.sign(data),
				},
				gidSeed: Buffer.from("a"),
				data: "entryA",
				next: [],
			});
			const two = await Entry.create({
				store,
				identity: {
					...signKey.keypair,
					sign: (data) => signKey.keypair.sign(data),
				},
				gidSeed: Buffer.from("a"),
				data: "entryB",
				next: [],
			});
			const three = await Entry.create({
				store,
				identity: {
					...signKey.keypair,
					sign: (data) => signKey.keypair.sign(data),
				},
				gidSeed: Buffer.from("a"),
				data: "entryC",
				next: [],
			});
			const log = new Log<string>();
			await log.open(store, {
				...signKey.keypair,
				sign: async (data: Uint8Array) => await signKey.keypair.sign(data),
			});
			await log.reset([one, two, three]);
			expect((await log.getHeads()).map((x) => x.hash)).toContainAllValues([
				one.hash,
				two.hash,
				three.hash,
			]);
		});
	});

	describe("toString", () => {
		let log: Log<string>;
		const expectedData =
			'"five"\n└─"four"\n  └─"three"\n    └─"two"\n      └─"one"';

		beforeEach(async () => {
			log = new Log<string>();
			await log.open(store, {
				...signKey.keypair,
				sign: async (data: Uint8Array) => await signKey.keypair.sign(data),
			});
			await log.append("one", { gidSeed: Buffer.from("a") });
			await log.append("two", { gidSeed: Buffer.from("a") });
			await log.append("three", { gidSeed: Buffer.from("a") });
			await log.append("four", { gidSeed: Buffer.from("a") });
			await log.append("five", { gidSeed: Buffer.from("a") });
		});

		it("returns a nicely formatted string", async () => {
			expect(await log.toString((p) => Buffer.from(p.data).toString())).toEqual(
				expectedData
			);
		});
	});

	describe("get", () => {
		let log: Log<any>;

		beforeEach(async () => {
			log = new Log<string>();
			await log.open(store, {
				...signKey.keypair,
				sign: async (data: Uint8Array) => await signKey.keypair.sign(data),
			});
			await log.append("one", {
				gidSeed: Buffer.from("a"),
				timestamp: new Timestamp({ wallTime: 0n, logical: 0 }),
			});
		});

		it("returns an Entry", async () => {
			const entry = await log.get((await log.toArray())[0].hash)!;
			expect(entry?.hash).toMatchSnapshot();
		});

		it("returns undefined when Entry is not in the log", async () => {
			const entry = await log.get(
				"zb2rhbnwihVVVVEGAPf9EwTZBsQz9fszCnM4Y8mJmBFgiyN7J"
			);
			assert.deepStrictEqual(entry, undefined);
		});
	});

	describe("setIdentity", () => {
		let log: Log<string>;

		beforeEach(async () => {
			log = new Log();
			await log.open(store, {
				...signKey.keypair,
				sign: async (data: Uint8Array) => await signKey.keypair.sign(data),
			});
			await log.append("one", { gidSeed: Buffer.from("a") });
		});

		it("changes identity", async () => {
			expect((await log.toArray())[0].metadata.clock.id).toEqual(
				signKey.keypair.publicKey.bytes
			);
			log.setIdentity({
				...signKey2.keypair,
				sign: signKey2.keypair.sign,
			});
			await log.append("two", { gidSeed: Buffer.from("a") });
			assert.deepStrictEqual(
				(await log.toArray())[1].metadata.clock.id,
				signKey2.keypair.publicKey.bytes
			);
			log.setIdentity({
				...signKey3.keypair,
				sign: signKey3.keypair.sign,
			});
			await log.append("three", { gidSeed: Buffer.from("a") });
			assert.deepStrictEqual(
				(await log.toArray())[2].metadata.clock.id,
				signKey3.keypair.publicKey.bytes
			);
		});
	});

	describe("has", () => {
		let log: Log<string>;

		beforeEach(async () => {
			log = new Log();
			await log.open(store, {
				...signKey.keypair,
				sign: async (data: Uint8Array) => await signKey.keypair.sign(data),
			});
			await log.append("one", { gidSeed: Buffer.from("a") });
		});

		it("returns true if it has an Entry", async () => {
			assert(log.has((await log.toArray())[0].hash));
		});

		it("returns true if it has an Entry, hash lookup", async () => {
			assert(log.has((await log.toArray())[0].hash));
		});

		it("returns false if it doesn't have the Entry", async () => {
			expect(
				await log.has("zb2rhbnwihVVVVEVVPf9EwTZBsQz9fszCnM4Y8mJmBFgiyN7J")
			).toEqual(false);
		});
	});

	describe("values", () => {
		it("returns all entries in the log", async () => {
			const log = new Log<string>();
			await log.open(store, {
				...signKey.keypair,
				sign: (data) => signKey.keypair.sign(data),
			});
			expect((await log.toArray()) instanceof Array).toEqual(true);
			expect(log.length).toEqual(0);
			await log.append("hello1");
			await log.append("hello2");
			await log.append("hello3");
			expect((await log.toArray()) instanceof Array).toEqual(true);
			expect(log.length).toEqual(3);
			expect((await log.toArray())[0].payload.getValue()).toEqual("hello1");
			expect((await log.toArray())[1].payload.getValue()).toEqual("hello2");
			expect((await log.toArray())[2].payload.getValue()).toEqual("hello3");
		});
	});
});
