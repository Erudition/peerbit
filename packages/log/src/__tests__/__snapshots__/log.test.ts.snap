// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Log get returns an Entry 1`] = `"zb2rhbvzcBVnVrHrnREeNygYYf5NybVPhmc9RTWf1xPgoW1iK"`;
